src/com/google/javascript/jscomp/NodeUtil.java

```diff
@@ -1461,89 +1461,95 @@ static boolean functionCallHasSideEffects(
+        case "acos":
+        case "acosh":
+        case "asin":
+        case "asinh":
+        case "atan":
+        case "atanh":
+        case "atan2":
+        case "cbrt":
+        case "ceil":
+        case "cos":
+        case "cosh":
+        case "exp":
+        case "expm1":
+        case "floor":
+        case "hypot":
+        case "log":
+        case "log10":
+        case "log1p":
+        case "log2":
+        case "max":
+        case "min":
+        case "pow":
+        case "round":
+        case "sign":
+        case "sin":
+        case "sinh":
+        case "sqrt":
+        case "tan":
+        case "tanh":
+        case "trunc":
           return false;
-        } else if (isTypedAsString(callee.getFirstChild(), compiler)) {
-          // Unlike regexs, string methods don't need to be hosted on a string literal
-          // to avoid leaking mutating global state changes, it is just necessary that
-          // the regex object can't be referenced.
-          String method = callee.getLastChild().getString();
-          Node param = callee.getNext();
-          if (param != null) {
-            if (param.isString()) {
-              if (STRING_REGEXP_METHODS.contains(method)) {
-                return false;
-              }
-            } else if (param.isRegExp()) {
-              if ("replace".equals(method)) {
-                // Assume anything but a string constant has side-effects
-                return !param.getNext().isString();
-              } else if (STRING_REGEXP_METHODS.contains(method)) {
-                return false;
-              }
+        case "random":
+          return !callNode.hasOneChild(); // no parameters
+        default:
+          // Unknown Math.* function, so fall out of this switch statement.
+      }
+    }
+
+    if (compiler != null && !compiler.hasRegExpGlobalReferences()) {
+      if (getPropNode.getFirstChild().isRegExp()
+          && REGEXP_METHODS.contains(getPropNode.getLastChild().getString())) {
+        return false;
+      } else if (isTypedAsString(getPropNode.getFirstChild(), compiler)) {
+        // Unlike regexs, string methods don't need to be hosted on a string literal
+        // to avoid leaking mutating global state changes, it is just necessary that
+        // the regex object can't be referenced.
+        String method = getPropNode.getLastChild().getString();
+        Node param = getPropNode.getNext();
+        if (param != null) {
+          if (param.isString()) {
+            if (STRING_REGEXP_METHODS.contains(method)) {
+              return false;
+            }
+          } else if (param.isRegExp()) {
+            if ("replace".equals(method)) {
+              // Assume anything but a string constant has side-effects
+              return !param.getNext().isString();
+            } else if (STRING_REGEXP_METHODS.contains(method)) {
+              return false;
             }
           }
         }
       }
     }
-
     return true;
```

Comment:
I suspect this `return true;`, assuming most property accesses have side effects, will block us from accepting this PR.

Right now @concavelenz is working on something with https://github.com/google/closure-compiler/blob/ab857cffd33d7114c58f47655d6f23795a27a610/src/com/google/javascript/jscomp/GatherGettersAndSetterProperties.java. This pass tracks which properties are potentially getters and setters.

I agree that the peephole optimization behavior is bad, but long-term we want to be more choosy about which property accesses we treat as potentially side-effectful. Does that sound reasonable?

Reply:
Yeah I agree this is why (just want to be more specific than _suspect_ for myself)

If you open https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/NodeUtil.java#L1463 and follow down then you can see a many `return false`, `return false` and it's never returning true. If the function didn't return in that `if` block, then it falls through to return true. Therefore I just moved the code to a separate function, and no change to the `functionCallHasSideEffects` at all because it's pure refactoring. 

Then, the real functional change is the addition of `case GETPROP` to the `nodeTypeMayHaveSideEffects` because it is the one that is called by the peephole optimiser. When it started to return true, I saw the failure in [the test](https://github.com/google/closure-compiler/blob/master/test/com/google/javascript/jscomp/PeepholeRemoveDeadCodeTest.java#L1289) that did `await some.thing()` [coming from](https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/PureFunctionIdentifier.java#L606) the `PureFunctionIdentifier` class, saying "unhandled side effect". Therefore, any function would now assume a side-effect exactly as you say. Despite the fact that there is an attempt to handle it by `setSideEffectsForGetProp`, it does not seem to be efficient: even the simple call `some.thing()` will now mark the function as having global side effects:

![Screenshot 2019-03-26 at 03 20 39](https://user-images.githubusercontent.com/21156791/54962502-2a8b0580-4f76-11e9-8160-b47eb79bf422.png)

I don't know why the function `setSideEffectsForGetProp` is not that efficient, and if it could be better, but it does not seem reasonable to mark a function tainted just because it accesses a property in an external object (although there's a chance as with getters). I do agree that side-effects in getters is a horrible thing to do, but there're some styles of programming that do exactly that, I think I saw that in some assertion libraries. Perhaps a note in the README or WIKI could be made about that? Plus as in with the issue, it was generated internally by the compiler -- is there any other places it does that. Finally, the fact that the green diff for `setSideEffectsForGetProp` finishes on line 666 makes me reluctant to continue (on top a number of failing tests) 😆 maybe when that change you pointed out earlier is implemented, it would be possible to improve the `setSideEffectsForGetProp` function. 

I also commented on the issue regarding if the latest changes help. ty

Reply:
would it still help to refactor the code into separate functions though?

Reply:
We have documented the getters with side effects issue here: https://github.com/google/closure-compiler/wiki/Compiler-Assumptions. I added another link in the readme under "Reporting a bug".

It's kind of sad that the compiler is the one creating getters with side-effects, yes :)